storage/src/adb/operation.rs

```diff
@@ -68,89 +61,48 @@ impl<K: Array, V: Array> Operation<K, V> {
         "array size too small for commit op"
     );
 
-    /// Create a new operation of the given type.
-    pub fn new(t: Type<K, V>) -> Self {
-        match t {
-            Type::Deleted(key) => Self::delete(key),
-            Type::Update(key, value) => Self::update(key, value),
-            Type::Commit(loc) => Self::commit(loc),
-        }
-    }
-
-    /// Create a new update operation that makes `key` have value `value`.
-    pub fn update(key: K, value: V) -> Self {
-        let mut data = Vec::with_capacity(Self::SIZE);
-        data.push(Self::UPDATE_CONTEXT);
-        data.extend_from_slice(&key);
-        data.extend_from_slice(&value);
-
-        Self {
-            data,
-            _phantom: std::marker::PhantomData,
-        }
-    }
-
-    /// Create a new delete operation that removes any value assigned to `key`.
-    pub fn delete(key: K) -> Self {
-        let mut data = Vec::with_capacity(Self::SIZE);
-        data.push(Self::DELETE_CONTEXT);
-        data.extend_from_slice(&key);
-        data.resize(Self::SIZE, 0);
-
-        Self {
-            data,
-            _phantom: std::marker::PhantomData,
-        }
-    }
-
-    /// Create a new commit operation that indicates the current floor on inactive operations is
-    /// `loc`.
-    pub fn commit(loc: u64) -> Self {
-        let mut data = Vec::with_capacity(Self::SIZE);
-        data.push(Self::COMMIT_CONTEXT);
-        data.extend_from_slice(&loc.to_be_bytes());
-        data.resize(Self::SIZE, 0);
-
-        Self {
-            data,
-            _phantom: std::marker::PhantomData,
-        }
-    }
-
-    pub fn to_key(&self) -> K {
-        K::decode(&self.data[1..K::SIZE + 1]).unwrap()
-    }
-
-    pub fn to_type(&self) -> Type<K, V> {
-        let key = K::decode(&self.data[1..K::SIZE + 1]).unwrap();
-        match self.data[0] {
-            Self::DELETE_CONTEXT => Type::Deleted(key),
-            Self::UPDATE_CONTEXT => {
-                let value = V::decode(&self.data[K::SIZE + 1..]).unwrap();
-                Type::Update(key, value)
-            }
-            Self::COMMIT_CONTEXT => {
-                let loc = u64::from_be_bytes(self.data[1..9].try_into().unwrap());
-                Type::Commit(loc)
-            }
-            _ => unreachable!(),
+    /// If this is a [Operation::Update] or [Operation::Deleted] operation, returns the key.
+    /// Otherwise, returns None.
+    pub fn to_key(&self) -> Option<K> {
```

Comment:
Since we've got the stuff pre-parsed, can we instead return Option<&K> ?

Reply:
I feel like this would run into annoying lifetime issues (?)

Reply:
couldn't you just clone it in that event?  I'm thinking of the typical usage pattern, which is to look up an Operation from a journal using the Index, and immediately check if its key matches the lookup key. You don't need the reference for long as you already have the copy you used for the lookup.

Reply:
Updated `to_key` and `to_value` to return references. As it stands, we never need to clone the returned references.